### PR TITLE
[WIP] Refactor core worker context

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Simplifies the code.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

**Comment by edoakes** [comment]
Depends on https://github.com/ray-project/ray/pull/6023.

File: src/ray/core_worker/core_worker.cc [hunk 1]
```diff
@@ -609,24 +597,62 @@ Status CoreWorker::ExecuteTask(const TaskSpecification &task_spec,
...
+      task_type, func, task_spec.GetRequiredResources().GetResourceMap(), args,
+      arg_reference_ids, return_ids, results);
+
+  ThreadContext::Get().ResetCurrentTaskId();
+  main_thread_task_id_ = TaskID::Nil();
+  if (task_spec.IsNormalTask()) {
+    current_job_id_ = JobID::Nil();
+    // Clear all actor handles at the end of each normal task.
+    for (const auto &handle : actor_handles_) {
+      RAY_CHECK_OK(gcs_client_.Actors().AsyncUnsubscribe(handle.first, nullptr));
```
Comment (zhijunfu) [inline src/ray/core_worker/core_worker.cc]: Could you please explain the logic here?

File: src/ray/core_worker/core_worker.cc [hunk 2]
```diff
@@ -609,24 +597,62 @@ Status CoreWorker::ExecuteTask(const TaskSpecification &task_spec,
...
     RAY_CHECK(return_ids.size() > 0);
     return_ids.pop_back();
+    RAY_CHECK(actor_id_.IsNil());
+    actor_id_ = task_spec.ActorCreationId();
     task_type = TaskType::ACTOR_CREATION_TASK;
-    SetActorId(task_spec.ActorCreationId());
-  } else if (task_spec.IsActorTask()) {
+  } else {
+    RAY_CHECK(current_job_id_ == task_spec.JobId());
+    RAY_CHECK(actor_id_ == task_spec.ActorId());
```
Comment (zhijunfu) [inline src/ray/core_worker/core_worker.cc]: the if-else-if-else block above is getting a bit too complicated. Please consider extract them to a separate function?

**Review by zhijunfu — CHANGES_REQUESTED** [review]
Thanks. Currently `WorkerContext` is a relatively thin wrapper of `WorkerThreadContext`, while also maintains some information itself. While it's probably good to remove it, on the other side, the current changes make the various variables scattered in core worker, which makes it harder to maintain and more prone to errors. How about we have a `WorkerInfo` structure just to wrap the` job_id, actor_id` stuff and the logic to initialize & update them so they are still self contained? CoreWorker can still call the interface of `WorkerInfo` instead of embedding all the logic. What do you think?

**Comment by edoakes** [comment]
@zhijunfu sorry this was a major WIP, should have labeled it. I am going to revisit this once a few outstanding PRs are in to avoid merge conflicts. Will consider the idea of a `WorkerInfo`, the logic for updating state is indeed getting a bit complex.

**Comment by zhijunfu** [comment]
@edoakes I see, thanks for the clarifications

**Pull request closed without merging**